src/core/eme/session.ts

```diff
@@ -314,174 +304,156 @@ function createSessionAndKeyRequest(
+export function createOrReuseSessionWithRetry(
+  encryptedEvent: MediaEncryptedEvent,
+  mediaKeysInfos: IMediaKeysInfos
+) {
+  return createOrReuseSession(
+    encryptedEvent,
+    mediaKeysInfos
+  ).catch((error) => {
+    if (error.code !== ErrorCodes.KEY_GENERATE_REQUEST_ERROR) {
+      throw error;
+    }
+    const firstLoadedSession = $loadedSessions.getFirst();
+    if (!firstLoadedSession) {
+      throw error;
+    }
+
+    log.warn("eme: could not create a new session, " +
+      "retry after closing a currently loaded session", error);
+
+    return $loadedSessions.deleteAndClose(firstLoadedSession)
+      .mergeMap(() => {
+        return createOrReuseSession(
+          encryptedEvent,
+          mediaKeysInfos
+        );
+      }
+      );
+  });
+}
+
+function createOrReuseSession(
+  encryptedEvent: MediaEncryptedEvent,
+  mediaKeysInfos: IMediaKeysInfos
+): Observable<ISessionEvent> {
+  // reuse currently loaded sessions without making a new key request
+  const initData = new Uint8Array(encryptedEvent.initData || new ArrayBuffer(0));
+  const initDataType = encryptedEvent.initDataType;
+
+  const loadedSession = $loadedSessions.get(initData);
+  if (loadedSession && loadedSession.sessionId) {
+    log.debug("eme: reuse loaded session", loadedSession.sessionId);
+    return Observable.of(createSessionEvent("reuse-session", loadedSession));
+  } else {
+    return createSession(mediaKeysInfos.mediaKeys, "temporary", initData)
+      .mergeMap((session) => {
+
+        const {
+          keySystem,
+          keySystemAccess,
+          mediaKeys,
+        } = mediaKeysInfos;
+        const mksConfig = keySystemAccess.getConfiguration();
+
+        let sessionType: MediaKeySessionType = "temporary"; // (default value)
+        const sessionTypes = mksConfig.sessionTypes;
+        const hasPersistence = (
+          sessionTypes && arrayIncludes(sessionTypes, "persistent-license"));
+
+        if (hasPersistence && keySystem.persistentLicense) {
+          sessionType = "persistent-license";
+
+          // if a persisted session exists in the store associated to this initData,
+          // we reuse it without a new license request through the `load` method.
+          const storedEntry = $storedSessions.get(initData);
+          if (storedEntry) {
+            return loadPersistentSession(storedEntry.sessionId, initData, session)
+              .catch(() => {
+                return createSession(mediaKeys, "persistent-license", initData)
+                  .mergeMap((_session) => {
+                    return generateKeyRequest(
+                      _session, initData, initDataType, sessionType);
+                  }).startWith(
+                    createSessionEvent(
+                      "loaded-session-failed",
+                      session,
+                      { storedSessionId: storedEntry.sessionId }
+                    ));
+              });
+          }
+        }
+        return generateRequestOnSession(
```

Comment:
In the ``if`` you use ``generateKeyRequest`` but here, you use ``generateRequestOnSession`` to do the same thing despite having the same variables available. That's not clean :broom: